src/seedu/addressbook/data/person/ReadOnlyPerson.java

```diff
@@ -37,50 +39,61 @@ default boolean isSameStateAs(ReadOnlyPerson other) {
      * Formats the person as text, showing all contact details.
      */
     default String getAsTextShowAll() {
-        final StringBuilder builder = new StringBuilder();
-        final String detailIsPrivate = "(private) ";
-        builder.append(getName())
-                .append(" Phone: ");
+        final Printable detailsIsPrivate = () -> "(private)";
+        List<Printable> details = new ArrayList<>();
+        details.add(getName());
         if (getPhone().isPrivate()) {
-            builder.append(detailIsPrivate);
+            details.add(detailsIsPrivate);
         }
-        builder.append(getPhone())
-                .append(" Email: ");
+        details.add(getPhone());
         if (getEmail().isPrivate()) {
-            builder.append(detailIsPrivate);
+            details.add(detailsIsPrivate);
         }
-        builder.append(getEmail())
-                .append(" Address: ");
+        details.add(getEmail());
         if (getAddress().isPrivate()) {
-            builder.append(detailIsPrivate);
+            details.add(detailsIsPrivate);
         }
-        builder.append(getAddress())
-                .append(" Tags: ");
+        details.add(getAddress());
+        final StringBuilder builder = new StringBuilder();
+        builder.append("Tags: ");
         for (Tag tag : getTags()) {
             builder.append(tag);
         }
-        return builder.toString();
+        return getPrintableString(details) + builder.toString();
     }
 
     /**
      * Formats a person as text, showing only non-private contact details.
      */
     default String getAsTextHidePrivate() {
-        final StringBuilder builder = new StringBuilder();
-        builder.append(getName());
+        List<Printable> nonPrivateDetails = new ArrayList<>();
+        nonPrivateDetails.add(getName());
         if (!getPhone().isPrivate()) {
-            builder.append(" Phone: ").append(getPhone());
+            nonPrivateDetails.add(getPhone());
         }
         if (!getEmail().isPrivate()) {
-            builder.append(" Email: ").append(getEmail());
+            nonPrivateDetails.add(getEmail());
         }
         if (!getAddress().isPrivate()) {
-            builder.append(" Address: ").append(getAddress());
+            nonPrivateDetails.add(getAddress());
         }
-        builder.append(" Tags: ");
+        final StringBuilder builder = new StringBuilder();
+        builder.append("Tags: ");
         for (Tag tag : getTags()) {
             builder.append(tag);
         }
-        return builder.toString();
+        return getPrintableString(nonPrivateDetails) + builder.toString();
+    }
+
+    /**
+     * Returns a concatenated version of the printable strings of each object.
+     */
+    private String getPrintableString(List<Printable> printables) {
+        String result = "";
+        for (Printable printable: printables) {
+            result += printable.getPrintableString();
```

Comment:
Usually in loops like this when the number of concatenations increases, StringBuilder is preferred, instead of using concatenating strings using `+`. Because Strings are immutable, what happens is each time you perform a `+`, the compiler internally builds a new StringBuilder to produce the concatenated string. For example, for `"abc", "def", "ghi"`, each concatenation generates a new string ie `"abcdef"` first before concatenating with `"ghi"`. This tends to become inefficient when concatenating a long list of strings. Appending all strings to StringBuilder and then converting to string will allow for compiler-level optimizations. Think about how you can improve what you have written.